### PR TITLE
PLT-1399 Write tests for all PIR passes that they preserve evaluation behaviour

### PR DESCRIPTION
**Pull request by thealmarty**
Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Changelog fragments have been written (if appropriate)
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting master unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-core/testlib/PlutusIR/Properties/Evaluation.hs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  -> Term TyName Name DefaultUni DefaultFun b
+  -> Either String (PLC.Term TyName Name DefaultUni DefaultFun ())
+evalNonPurePass pass biVariant pirTm = do
+  plcConfig <- PLC.getDefTypeCheckConfig (Original ())
+  lowered <- flip runReaderT (toDefaultCompilationCtx plcConfig) $
+    runQuoteT $ lowerTerm =<< pass (Original () <$ pirTm)
+  convertToEitherString $
+    evaluateCkNoEmit (defaultBuiltinsRuntimeForSemanticsVariant biVariant) (lowered $> ())
+
+-- | Generate builtin types. See this module's haddock.
```
Comment (thealmarty) [inline plutus-core/testlib/PlutusIR/Properties/Evaluation.hs:97]: This is how I generate the terms with `TyBuiltin` - but the generated terms included `Let` terms, as seen from the error from the deadcode test that I added. What am I doing wrong?

Comment (thealmarty) [reply plutus-core/testlib/PlutusIR/Properties/Evaluation.hs:119]: The error:
```
 Evaluation:                                                                   FAIL
          *** Failed! Error during compilation: Let bindings should have been eliminated before lowering(()) (after 2 tests):
          DefaultFunSemanticsVariant1
          DefaultFunSemanticsVariant2
          ty,tm = let
            ~f2_2 : data = B #
          in
          1
```

**Comment by thealmarty** [comment]
Added `caseReduce` test as an example pure test. All the rest of the tests should be straight forward to add. However, one still needs to figure out how to generate terms of builtin types only to avoid the `lowerTerm` errors. One may need to modify [genBuiltInTypeKind](https://github.com/input-output-hk/plutus/blob/da728215a9dcd5657d14ab4ad97765deb16db2e1/plutus-core/testlib/PlutusIR/Properties/Evaluation.hs#L110).

**Comment by kwxm** [comment]
Closing because stale.

**Pull request closed without merging**